### ISSUE TITLE
erroring when assoc is set to empty array

### ISSUE THREAD

**Issue by anasyb** [balderdashy/waterline#908]
With mongo, assume we have a user:
```javascript
posts: {
    collection: 'post',
    via: 'owner'
}
```
and obviously post:
```javascript
owner: {
    model: 'user'
}
```
Now let's create a user:
```javascript
User.create({id: 1, name: 'kitty', posts:[]}).exec(console.log);
```
This gives an unknown error:
```javascript
sails> Error (E_UNKNOWN) :: Encountered an unexpected error
: Error updating a record.
  at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:174:19
  at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:454:17
  at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:444:17
  at Array.forEach (native)
  at _each (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:46:24)
  at Object.taskComplete (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:443:13)
  at processImmediate [as _immediateCallback] (timers.js:345:15)

Details:  Error: Error updating a record.
```
This is totally unexpected since the default behavior of waterline is to return an empty array for associations. This means that if I retrieve a record with waterline and try to re-create it with the same waterline... it simply errors. 

Worth noting that the record was created successfully... so apart from the error msg I haven't noticed an issue.

**Comment by devinivy** [comment]
Sounds like a bug!  I will try to take a look at this today.  What version of waterline are you using?

**Comment by anasyb** [comment]
Am on waterline 0.10.18

**Comment by devinivy** [comment]
I tested on waterline 0.10.20 and 0.10.18 on sails-mysql and sails-memory for this situation in the case of a one-to-many and a many-to-many relationship.  Everything passed!  I have not tried mongo yet because I'm having trouble installing it on my machine.

**Comment by anasyb** [comment]
Thanks for the update...
I just tried to recreate the error and failed :( 
Seems like something fixed it... I should have kept better track of what went on. Sorry for the trouble!

**Comment by anasyb** [comment]
Ok got it...
**If I leave the `id` unspecified, everything goes thru.** If I specifiy the `id` and an empty association, it errors yet it create the `user` record. From the error I can see it says `Error updating a record` and I was telling my self: why does it say `updating` when I never asked it to update anything! I am guessing its trying to update a `post` record with the association and erroring there... just speculating.
So in a nutshell:

```javascript
//no errors
User.create({name: 'kitty', posts:[]}).exec(console.log);

//the error in the original post show up with this
User.create({id: '54e8cd422acb5e5a03d5c821', name: 'kitty', posts:[]}).exec(console.log);
```

**Comment by devinivy** [comment]
I cannot recreate this, sadly.  I've written a test here: https://github.com/devinivy/waterline-adapter-tests/commit/686769856a226b78d32eae7d85b3655d8084597e .  Let me know if it fails for you, or if there's some difference in our models causing the issue.

**Comment by anasyb** [comment]
I couldn't run the test, am not familiar with waterline tests... can you point me to some instructions on how to run waterline-adapter-tests?

I have an update: It turned out the array doesn't have to be empty! I tried creating nested records (which never happens with my app) and discovered something very interesting
```javascript
//this gives the same exact error!
User.create({id: '54e8cd422acb5e5a03d5c821', name: 'kitty', posts:[{title:"hi1"},{title:"hi2"}]}).exec(console.log);

//...checking db now shows that 1 User and 2 Posts are there

Post.findOne({title:'hi1'}).exec(console.log)
//both posts have their owner set to "owner" : "[object Object]", instead of '54e8cd422acb5e5a03d5c821'
```

This leads me to think it has to do more with sails-mongo. Its still puzzling that an empty array is also throwing the same error?!

**Comment by devinivy** [comment]
I'm unsure, but the error didn't appear in tests that Travis CI ran for mongo.

To run the tests for all community adapters: `npm test`
Tu run the test for a particular adapter: `node test/integration/runner.js {adapter-name}` (e.g. `node test/integration/runner.js sails-memory`)

Config for each of the adapters is in `test/integration/config`

**Comment by anasyb** [comment]
Couldn't run the tests... giving me some errors.

In any case, I've created a [mini-project](https://github.com/anasyb/waterline-issue-908), if you can run `sails console` and paste the following:

```javascript
//for some crazy commands from our alien commander let's create a profile before creating the users
Profile.create({id:'55340150298300641fe19c9e'}).exec(console.log)

//let's create our first pal!
User.create({id:'5534011a298300641fe19c98', profiles:['55340150298300641fe19c9e']}).exec(console.log)

//oh he's having a son, congrats to our beloved pal!
User.create({id:'55340150298300641fe19c9a', sonOf:'5534011a298300641fe19c98'}).exec(console.log)

//oh nooo.... the commander will be mad about this msg...
sails> Error (E_UNKNOWN) :: Encountered an unexpected error
: Error updating a record.
    at /Users/anas/.npm/lib/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:174:19
    at /Users/anas/.npm/lib/node_modules/sails/node_modules/async/lib/async.js:454:17
    at /Users/anas/.npm/lib/node_modules/sails/node_modules/async/lib/async.js:444:17
    at Array.forEach (native)
    at _each (/Users/anas/.npm/lib/node_modules/sails/node_modules/async/lib/async.js:46:24)
    at Object.taskComplete (/Users/anas/.npm/lib/node_modules/sails/node_modules/async/lib/async.js:443:13)
    at processImmediate [as _immediateCallback] (timers.js:345:15)

Details:  Error: Error updating a record.
```

Now checkout the mongo db for all crazy things:

```javascript
mongo> db.profile.find()
{
    "_id" : ObjectId("55340150298300641fe19c9e"),
    "createdAt" : ISODate("2015-04-19T19:55:41.867Z"),
    "updatedAt" : ISODate("2015-04-19T19:56:23.404Z"),
    "owner" : "[object Object]"
}

mongo> db.user.find()
{
    "_id" : ObjectId("5534011a298300641fe19c98"),
    "createdAt" : ISODate("2015-04-19T19:53:05.923Z"),
    "updatedAt" : ISODate("2015-04-19T19:53:05.923Z")
}
{
    "_id" : ObjectId("55340150298300641fe19c9a"),
    "createdAt" : ISODate("2015-04-19T19:53:19.301Z"),
    "updatedAt" : ISODate("2015-04-19T19:53:19.301Z")
}

mongo> db.user_sonof__user_fatherof.find()
{
    "_id" : ObjectId("553407af92684aee1fada54b"),
    "user_fatherOf" : ObjectId("5534011a298300641fe19c98"),
    "user_sonOf" : "[object Object]"
}
```

So you can see there are two relationships set to `[object Object]` instead of their real id. The first is in an association that's stored in a join tables, while the second one is with regards to an association that's being stored in as attribute of `Profile`.
Mind you though that if you try to create things from the other side (i.e. `Profile`) it works like a charms, could it be that this is what the tests are trying to do?!! 
```javascript
//this one pleases Mr. commander, it results in correct values in the db
Profile.create({id:'55340150298300641fe19c9e', owner:'5534011a298300641fe19c98'}).exec(console.log)
```

This bug is biting me hard, hope its not just me taking the bites :P

**Comment by devinivy** [comment]
@anasyb is it possible that those are ObjectId objects that haven't been converted to strings properly?

**Comment by anasyb** [comment]
I guess so... probably due to the same thing thats emitting the error msg... which I get in the linked project even with
```
User.create({id:'5534011a298300641fe19a98', profiles:[]}).exec(console.log)
```
Or maybe its the other way around: the lack of conversion of the BSON ObjectId is what's causing the error in the first place...

**Comment by anasyb** [comment]
Probably not... just changed the project to use sails disk, and am getting the same error msg when I run that line. Weird though: doesn't the disk adapter check of PK duplicates, I was able to insert it couple of time with the same `5534011a298300641fe19a98` ID?!

**Comment by devinivy** [comment]
@anasyb please give this branch of mine a try, and let me know how it goes.  It seems to fix the problem for me.

https://github.com/devinivy/sails-mongo/tree/patch-insert-ids

**Comment by devinivy** [comment]
The reason sails-disk isn't acting as expected has something to do with using mongo strings instead of integers.  If you use integers there it works.  I think there is a problem with key uniqueness in that adapter when you specify the key upon creation.

**Comment by devinivy** [comment]
https://github.com/balderdashy/sails-mongo/pull/262 merged!  This should be fixed, but if anyone experiences the issue again, we can reopen and continue the conversation.

**Comment by anasyb** [comment]
has been using this fix for a while now.. great work @devinivy 
what about the sails-disk thing, do you think its considered a bug and deserves an issue there?

**Comment by devinivy** [comment]
Yeah, I think it would be worthwhile creating an issue over there.
